quantlib/termstructures/volatility/equityfx/black_variance_surface.pyx

```diff
@@ -0,0 +1,113 @@
+from quantlib.time.calendar cimport Calendar
+from quantlib.time.daycounter cimport DayCounter
+from quantlib.time._daycounter cimport DayCounter as QlDayCounter
+
+from .black_vol_term_structure cimport BlackVarianceTermStructure
+from . cimport _black_variance_surface as _bvs
+from . cimport _black_vol_term_structure as _bvts
+from ..._vol_term_structure cimport VolatilityTermStructure
+
+cpdef public enum Extrapolation:
+    ConstantExtrapolation = _bvs.ConstantExtrapolation
+    InterpolatorDefaultExtrapolation = _bvs.InterpolatorDefaultExtrapolation
+
+cdef inline _bvs.BlackVarianceSurface* get_bvs(BlackVarianceSurface bvs):
+    """ Utility function to extract a properly casted BlackVarianceSurface out
+    of the internal _thisptr attribute of the BlackVolTermStructure base class.
+    """
+
+    cdef _bvs.BlackVarianceSurface* ref = <_bvs.BlackVarianceSurface*>bvs._thisptr.get()
+    return ref
+
+
+cdef class BlackVarianceSurface(BlackVarianceTermStructure):
+    """"
+    Black volatility surface modelled as variance surface
+    This class calculates time/strike dependent Black volatilities
+    using as input a matrix of Black volatilities observed in the
+    market.
+
+    The calculation is performed interpolating on the variance surface.
+    Bilinear interpolation is used as default; this can be changed by the
+    setInterpolation() method.
+
+    todo check time extrapolation
+
+    Attributes
+    ----------
+    reference_date : Date
+    cal : Calendar
+    dates : list of `obj`:Date
+    strikes : list of floats
+    black_vol_matrix : Matrix
+    dc : DayCounter
+    lower_extrap : Extrapolation
+    upper_extrap : Extrapolation
+
+    """
+
+    def __init__(self,
+                 Date reference_date,
+                 Calendar cal,
+                 list dates,
+                 vector[Real] strikes,
+                 Matrix black_vol_matrix,
+                 DayCounter dc,
+                 Extrapolation lower_extrap = Extrapolation.InterpolatorDefaultExtrapolation,
+                 Extrapolation upper_extrap = Extrapolation.InterpolatorDefaultExtrapolation,
+                 ):
+
+        cdef vector[_Date] _dates
+        for d in dates:
+            _dates.push_back(deref((<Date?>d)._thisptr))
+
+        self._thisptr = shared_ptr[VolatilityTermStructure](
+                                new _bvs.BlackVarianceSurface(
+                                             deref(reference_date._thisptr),
+                                             deref(cal._thisptr),
+                                             _dates,
+                                             strikes,
+                                             (black_vol_matrix)._thisptr,
+                                             deref(dc._thisptr),
+                                             lower_extrap,
+                                             upper_extrap,
+                                             ))
+
+    # TermStructure interface
```

Comment:
Commented code should be removed or uncommented. Cany you check?

Reply:
removed in 1244cdc